### PR TITLE
Add Rotary Encoder component support

### LINKED ISSUE #392: Add Rotary Encoder component support
# Description

Add Rotary Encoder component support 

Rotary Encoder

I'm adapting the [pigpio code](http://abyz.co.uk/rpi/pigpio/examples.html#Python_rotary_encoder_py). When I finished, I will comment this Issue for Code Review.
## References
- https://www.youtube.com/watch?v=cn83jR2mchw
- http://playground.arduino.cc/Main/RotaryEncoders
- http://www.pjrc.com/teensy/td_libs_Encoder.html
- http://abyz.co.uk/rpi/pigpio/examples.html#Python_rotary_encoder_py

### PR DESCRIPTION
**Pull request by SrMouraSilva**
Fix RotaryEncoder support based in @lurch suggestions (https://github.com/PedalPi/Physical/issues/1)

---
This change is [Reviewable](https://reviewable.io/reviews/rpi-distro/python-gpiozero/482)

### CODE REVIEW

**Comment by bennuttall** [comment]
Comprehensive PR! Thanks.

I've a few comments to make and I'll want @lurch and @waveform80 to check over.

File: gpiozero/__init__.py [hunk 1]
```diff
@@ -132,3 +132,7 @@
     CPUTemperature,
     TimeOfDay,
 )
+from .rotary_encoder import (
```
Comment (bennuttall) [inline gpiozero/__init__.py]: I think these classes belong in boards.py, not in their own file.

File: gpiozero/rotary_encoder.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
+# -*- coding: utf-8 -*-
+from gpiozero import Device, DigitalInputDevice, Button, CompositeDevice
+
+
+class RotaryEncoder(Device):
```
Comment (bennuttall) [inline gpiozero/rotary_encoder.py]: Extending Device is quite low-level. I guess it doesn't get any source/values stuff?

File: gpiozero/rotary_encoder.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+    3V3 pin, then set *pull_up* to ``False`` in the :class:`RotaryEncoder` constructor.
+
+    The following example will print a Rotary Encoder change direction::
+
+        from gpiozero import RotaryEncoder
+
+        def change(value):
+            if value > 0:
+                print("clockwise")
+
```
Comment (bennuttall) [inline gpiozero/rotary_encoder.py]: Remove this line space

File: gpiozero/rotary_encoder.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        for valid pin numbers.
+
+    :param bool pull_up:
+        If ``True`` (the default), the GPIO pins will be pulled high by default.
+        In this case, connect the middle GPIO pin to ground. If ``False``, 
+        the GPIO pins will be pulled low by default. In this case,
+        connect the middle pin of the RotaryEncoder to 3V3.
+    """
+
+    when_rotated = lambda *args: None
```
Comment (bennuttall) [inline gpiozero/rotary_encoder.py]: This is a class property, isn't it? Should it not be self.when_rotated defined within the __init__ method?

File: gpiozero/rotary_encoder.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+    @property
+    def value(self):
+        return None
+
+    def __repr__(self):
+        return "<gpiozero.%s object on pin_a %r, pin_b %r, pull_up=%s, is_active=%s>" % (
+                self.__class__.__name__, self.pin_a.pin, self.pin_b.pin, self.pin_a.pull_up, self.is_active)
+
+
+class TableValues:
```
Comment (bennuttall) [inline gpiozero/rotary_encoder.py]: Should this be `TableValues(object)` for Python 2?

File: gpiozero/rotary_encoder.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    @staticmethod
+    def value(new_b_value, new_a_value, old_b_value, old_a_value):
+        index = TableValues.calcule_index(new_b_value, new_a_value, old_b_value, old_a_value)
+        try:
+            return TableValues.values[index]
+        except KeyError:
+            return 0
+
+    @staticmethod
+    def calcule_index(new_b_value, new_a_value, old_b_value, old_a_value):
```
Comment (bennuttall) [inline gpiozero/rotary_encoder.py]: Should this be `calculate_index`?

File: gpiozero/rotary_encoder.py [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+            value += 4
+        if old_b_value:
+            value += 2
+        if old_a_value:
+            value += 1
+
+        return value
+
+
+class RotaryEncoderWithButton(CompositeDevice):
```
Comment (bennuttall) [inline gpiozero/rotary_encoder.py]: Not sure about the `With` in this class name.

Comment (bennuttall) [inline gpiozero/rotary_encoder.py]: Also what about values?

File: gpiozero/rotary_encoder.py (hunk 2 above)
Comment (SrMouraSilva) [inline gpiozero/rotary_encoder.py]: The initial code did not extends, but I thought it was important to support the ``close`` method. Suggestion for give [close](http://gpiozero.readthedocs.io/en/v1.3.1/api_boards.html#gpiozero.CompositeDevice.close) support?

File: gpiozero/rotary_encoder.py (hunk 5 above)
Comment (SrMouraSilva) [inline gpiozero/rotary_encoder.py]: I did not understand. Python 2.7 works https://travis-ci.org/RPi-Distro/python-gpiozero/jobs/165127085

File: gpiozero/rotary_encoder.py (hunk 6 above)
Comment (SrMouraSilva) [inline gpiozero/rotary_encoder.py]: Ops! Yes. A portuguese wrong translation ("**Calcule** o índice" === "Calculate the index")

File: gpiozero/rotary_encoder.py (hunk 7 above)
Comment (SrMouraSilva) [inline gpiozero/rotary_encoder.py]: > Not sure about the With in this class name.
 
``RotaryEncoderClickable``?

> Also what about values?

[Same as quoted in here](https://github.com/RPi-Distro/python-gpiozero/pull/482#discussion_r82022793)

**Comment by SrMouraSilva** [comment]
Update code to gpiozero 1.4.x

**Comment by SrMouraSilva** [comment]
If corrections are required, please inform.

**Comment by SrMouraSilva** [comment]
Hello, @bennuttall. 

I saw that it was removed from version 1.5. Is there any programming for launching this functionality? Do you need anything? Documentation, sample codes?

**Comment by odiroot** [comment]
Hi, did you have problems with your encoder being "chatty"? 
I tried modified your code a bit to support lower level SW de-bounce:
```
    def __init__(self, pin_a, pin_b, pull_up=True, bounce_time=None):
        self.when_rotated = lambda *args: None
        self.pin_a = DigitalInputDevice(pin=pin_a, pull_up=pull_up, bounce_time=bounce_time)
        self.pin_b = DigitalInputDevice(pin=pin_b, pull_up=pull_up, bounce_time=bounce_time)
```
Still it helps just a bit.

**Comment by SrMouraSilva** [comment]
Hello, @odiroot!

I had no problems. Maybe your encoder is experiencing a problem, maybe the power supply is in trouble.
Anyway, you did a great addition. I'll add this to merge request.

A little note. I use the rotary encoder "pure" without any plate. Rotary encodes sold as shields have some components that I do not know what they are for. Maybe they are to avoid just this problem you are facing!

See "Filtering Out the Contact Noise of a Mechanical Encoder":
https://www.allaboutcircuits.com/projects/how-to-use-a-rotary-encoder-in-a-mcu-based-project/

image

**Comment by odiroot** [comment]
Hi @SrMouraSilva 
Thanks for quick answer. I think my encoder may be slightly different. Also different from one in that link.
I believe this is my part: https://www.digikey.com/product-detail/en/panasonic-electronic-components/EVE-PDBRL408B/P12421-ND/637158

I actually noticed something funny, a sort of a shortcut.

1. Listen on both raising and falling edges on two pins (A & B) (excuse a RPi.GPIO example):
```python
GPIO.add_event_detect(leg_a, GPIO.BOTH, callback=on_rotate)
GPIO.add_event_detect(leg_b, GPIO.BOTH, callback=on_rotate)
```
2. Take a note (store state in some variable) of last direction being triggered. 
3. Use the same handler (`on_rotate` above) to handle both inputs and ignore the trigger if it's from the same pin/leg as before:
```python
def on_rotate(channel):
    global previous
    if channel == previous:  # Ignore
        return
    else:
        previous = channel  # Note
    # ...
```
4. Actually don't do anything more (than above) if the triggering leg/pin is A (A is the leading signal).
5. In case signal A and signal B are equal (both `0` or both `1`) assume CCW rotation.
6. In case signal A are unequal (one is `0` and the other is `1`) assume CW rotation.

Somehow it works super well for me even though I throw away half of the triggers. Again I went with raw _RPi.GPIO_ instead of _gpizero_ (didn't have time to adapt the code to class-based).

**Comment by mankyd** [comment]
Is there any update on this? I could use this at the moment. 2 years ago it was mentioned that it was going to be merged, but it still seems to be missing from the available io devices.

**Comment by SrMouraSilva** [comment]
@mankyd, in fact the merge hasn't been done yet.
@bennuttall, when you are going to start the merge process, please mention me here for me to resolve the conflicts.

**Comment by tomhaines** [comment]
Is there a plan to merge this? It would be very helpful to have rotary encoder support

**Comment by yurikleb** [comment]
Is there an estimation on if and when would this be merged? would be very useful to have a rotary encoder class available.

In the meanwhile, what would be the best way to use Gpiozero to interface with a 2 (or more) pins device (such as an encoder)?

Based on the documentation the "[GPIODevice](https://gpiozero.readthedocs.io/en/stable/api_input.html#gpiodevice)" base class "provides the services common to all **single-pin** GPIO devices"

**Comment by bennuttall** [comment]
Sorry I think it's unlikely we'd merge this any time soon, working on other things.

You can use [CompositeDevice](https://gpiozero.readthedocs.io/en/stable/api_boards.html#gpiozero.CompositeDevice) to create a class with two devices. There are plenty of examples of this - see the class hierarchy here: https://gpiozero.readthedocs.io/en/stable/api_generic.html

**Comment by waveform80** [comment]
Sorry this has taken so long to get to, and many thanks for the PR. As Ben noted it's pretty comprehensive and it's great to see tests being included too. I've had a play with this branch on a couple of basic rotary encoders and, as noted in some of the comments above, there's a few issues with jittery encoders. After a little research I managed to work around that by replacing the event handlers with a simple state machine that permits a certain amount of back'n'forth once the initial direction has been established by the first edge that arrives.

By that point things were getting a bit messy and I started a fresh branch to build upon so I'm afraid I'm going to close this PR, in favour of a fresh one with the replaced code but I just wanted to say thank you for the contribution: we probably wouldn't have got to rotary encoders without it (there's so much else on my todo list, and having a good first step like this really helps in deciding what to deal with next). Anyway, I've opened #928 as the "spiritual continuation" of this PR and hopefully we can get this included shortly!

**Pull request closed without merging**